File: israeli-id-validation.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

bool isValidIsraeliID(std::string id) {
    // Add leading zeroes if ID is less than 9 digits
    id.insert(0, 9 - id.length(), '0');

    // Convert ID to array of integers
    std::vector<int> idArr(id.length());
    std::transform(id.begin(), id.end(), idArr.begin(), [](unsigned char c) { return c - '0'; });

    // Multiply first 8 digits by check array
    int checkArr[] = {1, 2, 1, 2, 1, 2, 1, 2};
    std::vector<int> calcArr(8);
    for (int i = 0; i < 8; i++) {
        calcArr[i] = idArr[i] * checkArr[i];
    }

    // Calculate sum of resulting array
    int sum = 0;
    for (int i = 0; i < 8; i++) {
        sum += calcArr[i] > 9 ? calcArr[i] - 9 : calcArr[i];
    }

    // Calculate check digit
    int checkDigit = (10 - (sum % 10)) % 10;

    // Compare check digit with last digit of ID
    return idArr[8] == checkDigit;
}

File: israeli-id-validation.rs
fn is_valid_israeli_id(id: &str) -> bool {
    // Add leading zeroes if ID is less than 9 digits
    let id = format!("{:09}", id);

    // Convert ID to array of integers
    let id_arr: Vec<i32> = id.chars().map(|c| c.to_digit(10).unwrap() as i32).collect();

    // Multiply first 8 digits by check array
    let check_arr = [1, 2, 1, 2, 1, 2, 1, 2];
    let mut calc_arr = vec![0; 8];
    for i in 0..8 {
        calc_arr[i] = id_arr[i] * check_arr[i];
    }

    // Calculate sum of resulting array
    let sum = calc_arr.iter().map(|&val| if val > 9 { val - 9 } else { val }).sum();

    // Calculate check digit
    let check_digit = (10 - (sum % 10)) % 10;

    // Compare check digit with last digit of ID
    id_arr[8] == check_digit
}
